### PR TITLE
Add Prompt Tuning

### PR DESCRIPTION
**Pull request by lenglaender**
Depends on: #591 

This PR adds Prompt Tuning (https://aclanthology.org/2021.emnlp-main.243/)

Currently the Prompt Tuning Layer was created and the BERT model has been adapted

### CODE REVIEW

**Review by calpt — COMMENTED** [review]
did a first review pass, looks good overall, haven't checked all details though

**Comment by calpt** [comment]
Please also add row in table on main readme

**Comment by raphaelreimann** [comment]
Hey, love the integration of prompt tuning! I would love to use it when it is stable, do you know when this will be merged? 🙂 @lenglaender @calpt

**Comment by lenglaender** [comment]
@raphaelreimann We're currently fixing the last few tests that aren't running. For every model except CLIP, EncoderDecoder, GPT-2, GPT-J, LLaMA and T5 the implementation is already working.

As soon as this PR and the sync PR (#602) are merged, we will release the new Adapters library (#584).

**Comment by raphaelreimann** [comment]
@lenglaender Thank you for the update on this, great work!

File: src/adapters/model_mixin.py [hunk 1]
```diff
@@ -40,22 +41,6 @@ def init_adapters(self, model_config, adapters_config, **kwargs):
         if hasattr(super(), "init_adapters"):
             super().init_adapters(self.config, self.adapters_config, **kwargs)
 
-        self.hook_after_embeddings(self._hook_fn)
-
-    def _hook_fn(self, module, args, output):
-        new_output = self.invertible_adapters_forward(output)
-        return new_output
-
-    def hook_after_embeddings(self, hook_fn: Callable):
```
Comment (hSterz) [inline src/adapters/model_mixin.py:49]: Why are we deleting/refactoring this?

File: tests_adapters/methods/test_prompt_tuning.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        model = self.get_model()
+        self.run_forward_test(model, PromptTuningConfig(prompt_length=10))
+
+    def test_load_prompt_tuning(self):
+        self.run_load_test(PromptTuningConfig(prompt_length=10))
+
+    def test_load_full_model_prefix_tuning(self):
+        self.run_full_model_load_test(PromptTuningConfig(prompt_length=10))
+
+    def test_train_prefix_tuning(self):
```
Comment (hSterz) [inline tests_adapters/methods/test_prompt_tuning.py:41]: Probably, a left over from copying over tests but I guess this should be called `test_train_prompt_tuning`

**Review by hSterz — APPROVED** [review]
Looks good to me just a few minor comments

File: src/adapters/model_mixin.py (hunk 1 above)
Comment (lenglaender) [reply src/adapters/model_mixin.py:49]: We now have 2 adapter methods that are hooked in after the embeddings (invertible adapter & prompt tuning) and have therefore decided to put them in the new `post_embedding_forward` function of the `ModelBaseAdaptersMixin`. This makes the implementation cleaner and it is easier to add additional adapter methods later, which need to be added after the embedding layer as well.

File: tests_adapters/methods/test_prompt_tuning.py (hunk 2 above)
Comment (lenglaender) [reply tests_adapters/methods/test_prompt_tuning.py]: Yes, thanks for pointing that out! I've renamed it

**Comment by lenglaender** [comment]
For the subsequent PR, which will extend Prompt Tuning with composition support and which will add support for the not yet supported models, we should think about realising the change of the attention mask for prefix tuning within the function `prefix_attention_mask`.
